contracts/strategies/donation-voting-merkle-distribution-direct-transfer/DonationVotingMerkleDistributionDirectTransferStrategy.sol

```diff
@@ -70,9 +72,48 @@ contract DonationVotingMerkleDistributionDirectTransferStrategy is DonationVotin
                 // the EIP712 hash of `_permit`.
                 p2Data.signature
             );
+        } else if (permitType == PermitType.Permit) {
+            (bytes32 r, bytes32 s, uint8 v) = splitSignature(p2Data.signature);
+            // The tx can be front-run, and another user can use the permit message and signature to invalidate the nonce.
+            // In this case the permit call will fail, but it means that the contract already has allowance for the token.
+            try IERC20Permit(token).permit(_sender, address(this), amount, p2Data.permit.deadline, v, r, s) {}
+            catch Error(string memory reason) {
+                if (IERC20(token).allowance(msg.sender, address(this)) < amount) {
+                    revert(reason);
+                }
+            } catch (bytes memory reason) {
+                if (IERC20(token).allowance(msg.sender, address(this)) < amount) {
+                    revert(string(reason));
+                }
+            }
+            IERC20(token).transferFrom(_sender, _recipients[recipientId].recipientAddress, amount);
```

Comment:
On etherscan the transaction will be marked as Failed if the catch is entered. this might cause a lot of confusion.

wdyt about implementing it like that:

```
IERC20 t = IERC20(token);
if (t.allowance(msg.sender, address(this)) < amount) {
    t.permit(_sender, address(this), amount, p2Data.permit.deadline, v, r, s);
}
t.transferFrom(_sender, _recipients[recipientId].recipientAddress, amount);
 ```

Reply:
great point @0xKurt 

we have 2 options. 

1 - the current implementation is checking the allowance only when the call to `permit` fails, and it can fail only in very rare cases when the signed permit expired. It can happen front running the tx or using the same nonce before the tx is mined. 

2 - the other option you proposed is to always check the allowance and call `permit` only when it's needed. 

Given that this problem is so rare, I would like to go with option 1, so that we never use more gas then needed calling the other contract,  unless the very rare case happens.

Reply:
lets keep the  current implementation